storm-core/src/jvm/org/apache/storm/daemon/supervisor/SyncSupervisorEvent.java

```diff
@@ -0,0 +1,632 @@
+            for (Map.Entry<Integer, LocalAssignment> entry : newAssignment.entrySet()) {
+                convertNewAssignment.put(entry.getKey().longValue(), entry.getValue());
+            }
+            supervisorData.setCurrAssignment(convertNewAssignment);
+            // remove any downloaded code that's no longer assigned or active
+            // important that this happens after setting the local assignment so that
+            // synchronize-supervisor doesn't try to launch workers for which the
+            // resources don't exist
+            if (Utils.isOnWindows()) {
+                shutdownDisallowedWorkers();
+            }
+            for (String stormId : allDownloadedTopologyIds) {
+                if (!stormcodeMap.containsKey(stormId)) {
+                    LOG.info("Removing code for storm id {}.", stormId);
+                    rmTopoFiles(conf, stormId, supervisorData.getLocalizer(), true);
+                }
+            }
+            syncProcessManager.add(syncProcesses);
+        } catch (Exception e) {
+            LOG.error("Failed to Sync Supervisor", e);
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    private void killExistingWorkersWithChangeInComponents(SupervisorData supervisorData, Map<Integer, LocalAssignment> existingAssignment,
+            Map<Integer, LocalAssignment> newAssignment) throws Exception {
+        LocalState localState = supervisorData.getLocalState();
+        Map<Integer, LocalAssignment> assignedExecutors = localState.getLocalAssignmentsMap();
+        if (assignedExecutors == null) {
+            assignedExecutors = new HashMap<>();
+        }
+        int now = Time.currentTimeSecs();
+        Map<String, StateHeartbeat> workerIdHbstate = syncProcesses.getLocalWorkerStats(supervisorData, assignedExecutors, now);
+        Map<Integer, String> vaildPortToWorkerIds = new HashMap<>();
+        for (Map.Entry<String, StateHeartbeat> entry : workerIdHbstate.entrySet()) {
+            String workerId = entry.getKey();
+            StateHeartbeat stateHeartbeat = entry.getValue();
+            if (stateHeartbeat != null && stateHeartbeat.getState() == State.VALID) {
+                vaildPortToWorkerIds.put(stateHeartbeat.getHeartbeat().get_port(), workerId);
+            }
+        }
+
+        Map<Integer, LocalAssignment> intersectAssignment = new HashMap<>();
+        for (Map.Entry<Integer, LocalAssignment> entry : newAssignment.entrySet()) {
+            Integer port = entry.getKey();
+            if (existingAssignment.containsKey(port)) {
+                intersectAssignment.put(port, entry.getValue());
+            }
+        }
+
+        for (Integer port : intersectAssignment.keySet()) {
+            List<ExecutorInfo> existExecutors = existingAssignment.get(port).get_executors();
+            List<ExecutorInfo> newExecutors = newAssignment.get(port).get_executors();
+            if (newExecutors.size() != existExecutors.size()) {
```

Comment:
@jerrypeng  I may not follow up your means. If they contain the same elements  (without considering ordering) , it don't  shut down the worker.
···
           for (ExecutorInfo executorInfo : newExecutors) {
                if (!existExecutors.contains(executorInfo)) {
                    syncProcesses.shutWorker(supervisorData, supervisorData.getWorkerManager(), vaildPortToWorkerIds.get(port));
                    break;
                }
            }
···